src/Control/Newtype/Generic.hs

```diff
@@ -0,0 +1,203 @@
+--
+-- The reason for the signature of the /hof/ is due to 'ala' not caring about structure.
+-- To illustrate why this is important, consider this alternative implementation of 'under2':
+--
+-- > under2 :: (Newtype n o, Newtype n' o')
+-- >        => (o -> n) -> (n -> n -> n') -> (o -> o -> o')
+-- > under2' pa f o0 o1 = ala pa (\p -> uncurry f . bimap p p) (o0, o1)
+--
+-- Being handed the "packer", the /hof/ may apply it in any structure of its choosing –
+-- in this case a tuple.
+--
+-- >>> ala Sum foldMap [1,2,3,4]
+-- 10
+ala :: (Coercible n o, Newtype n' o')
+    => (o `to` n) -> ((o -> n) -> b -> n') -> (b -> o')
+ala pa hof = ala' pa hof id
+
+-- | This is the original function seen in Conor McBride's work.
+-- The way it differs from the 'ala' function in this package,
+-- is that it provides an extra hook into the \"packer\" passed to the hof.
+-- However, this normally ends up being @id@, so 'ala' wraps this function and
+-- passes @id@ as the final parameter by default.
+-- If you want the convenience of being able to hook right into the hof,
+-- you may use this function.
+--
+-- >>> ala' Sum foldMap length ["hello", "world"]
+-- 10
+--
+-- >>> ala' First foldMap (readMaybe @Int) ["x", "42", "1"]
+-- Just 42
+ala' :: (Coercible n o, Newtype n' o')
+     => (o `to` n) -> ((a -> n) -> b -> n') -> (a -> o) -> (b -> o')
+--ala' _ hof f = unpack . hof (pack . f)
+ala' _ = coerce
+
+-- | A very simple operation involving running the function \'under\' the newtype.
+--
+-- >>> under Product (stimes 3) 3
+-- 27
+under :: (Coercible n o, Newtype n' o')
+      => (o `to` n) -> (n -> n') -> (o -> o')
+--under _ f = unpack . f . pack
+under _ = coerce
+
+-- | The opposite of 'under'. I.e., take a function which works on the
+-- underlying types, and switch it to a function that works on the newtypes.
+--
+-- >>> over All not (All False)
+-- All {getAll = True}
+over :: (Coercible n o,  Newtype n' o')
+     => (o `to` n) -> (o -> o') -> (n -> n')
+--over _ f = pack . f . unpack
+over _ = coerce'
+
+-- | Lower a binary function to operate on the underlying values.
+--
+-- >>> under2 Any (<>) True False
+-- True
+--
+-- @since 0.5.2
+under2 :: (Coercible n o, Newtype n' o')
+       => (o `to` n) -> (n -> n -> n') -> (o -> o -> o')
+--under2 _ f o0 o1 = unpack $ f (pack o0) (pack o1)
+under2 _ = coerce
+
+-- | The opposite of 'under2'.
+--
+-- @since 0.5.2
+over2 :: (Coercible n o, Newtype n' o')
+       => (o `to` n) -> (o -> o -> o') -> (n -> n -> n')
+--over2 _ f n0 n1 = pack $ f (unpack n0) (unpack n1)
+over2 _ = coerce'
+
+-- | 'under' lifted into a Functor.
+underF :: (Coercible n o, Newtype n' o', Functor f, Functor g)
+       => (o `to` n) -> (f n -> g n') -> (f o -> g o')
+underF _ f = fmap unpack . f . fmap coerce
+
+-- | 'over' lifted into a Functor.
+overF :: (Coercible n o, Newtype n' o', Functor f, Functor g)
+      => (o `to` n) -> (f o -> g o') -> (f n -> g n')
+overF _ f = fmap pack . f . fmap coerce
```

Comment:
Should these use `QuantifiedConstraints` or (for older GHC versions) `Data.Constraint.Forall`? That will limit them to representational functors, but on the plus side they'll be guaranteed efficient and they'll work for contravariant and invariant functors as well.

Reply:
Oh, actually, we just need `FlexibleContexts`, don't we?